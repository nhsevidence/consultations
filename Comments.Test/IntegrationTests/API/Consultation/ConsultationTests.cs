using System;
using System.Linq;
using System.Net;
using System.Threading.Tasks;
using Comments.Services;
using Comments.Test.Infrastructure;
using Comments.ViewModels;
using Newtonsoft.Json;
using NICE.Feeds.Tests.Infrastructure;
using Shouldly;
using Xunit;
using TestBase = Comments.Test.Infrastructure.TestBase;

namespace Comments.Test.IntegrationTests.API.Consultation
{
    public class ConsultationTests : TestBase
    {
        public ConsultationTests() : base(Feed.ConsultationCommentsPublishedDetailMulitpleDoc, true, Guid.Empty, "Benjamin Button") { }

        [Theory]
        [InlineData(0)]
        [InlineData(-1)]
        public async Task Get_Consultation_Feed_Using_Invalid_Consultation_Id_Throws_error(int consultationId)
        {
	        //Act
	        var response = await _client.GetAsync($"/consultations/api/Consultation?consultationId={consultationId}");
			var responseString = await response.Content.ReadAsStringAsync();

	        //Assert
	        response.StatusCode.ShouldBe(HttpStatusCode.InternalServerError);
	        responseString.ShouldMatchApproved(new Func<string, string>[] { Scrubbers.ScrubErrorMessage });
		}

        [Fact]
        public async Task Get_Consultation_Feed_Returns_Populated_Feed()
        {
            //Arrange (in base constructor)

            // Act
            var response = await _client.GetAsync("/consultations/api/Consultation?consultationId=1");
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            responseString.ShouldMatchApproved();
        }

	    [Fact]
	    public async Task Get_Draft_Consultation_Feed_Returns_Populated_Feed()
	    {
		    //Arrange (in base constructor)

		    // Act
		    var response = await _client.GetAsync("/consultations/api/DraftConsultation?consultationId=113&documentId=1&reference=GID-NG10186");
		    response.EnsureSuccessStatusCode();
		    var responseString = await response.Content.ReadAsStringAsync();

		    // Assert
		    responseString.ShouldMatchApproved();
	    }

	    [Fact]
	    public async void CommentsQuestionsAndAnswers_CanBeRead()
	    {
		    // Arrange
		    ResetDatabase();
			_context.Database.EnsureCreated();

			var sourceURI = "consultations://./consultation/1/document/1/chapter/introduction";
		    var commentText = Guid.NewGuid().ToString();
		    var questionText = Guid.NewGuid().ToString();
		    var answerText = Guid.NewGuid().ToString();
		    var createdByUserId = Guid.Empty;
		    var authenticateService = new FakeAuthenticateService(authenticated: true);

		    AddCommentsAndQuestionsAndAnswers(sourceURI, commentText, questionText, answerText, createdByUserId, (int)StatusName.Draft, _context);
		    //var submitService = new SubmitService(_context, _fakeUserService, _consultationService);
		    var commentService = new CommentService(_context, _fakeUserService, authenticateService, _consultationService);

		    // Act    
		    var viewModel = commentService.GetCommentsAndQuestions(sourceURI);

		    //Assert
		    viewModel.Comments.Single().CommentText.ShouldBe(commentText);
		    var question = viewModel.Questions.Single();
		    question.QuestionText.ShouldBe(questionText);
		    question.Answers.Single().AnswerText.ShouldBe(answerText);


		    //sample code:

		    var resp = await _client.GetAsync($"/consultations/api/Comments?sourceURI={WebUtility.UrlEncode(sourceURI)}");
		    resp.EnsureSuccessStatusCode();
		    var responseString = await resp.Content.ReadAsStringAsync();
		    var anotherQuestion = JsonConvert.DeserializeObject<CommentsAndQuestions>(responseString);
		    anotherQuestion.Questions.Single().Answers.Single().AnswerText.ShouldBe(answerText);

	    }
	}
}
