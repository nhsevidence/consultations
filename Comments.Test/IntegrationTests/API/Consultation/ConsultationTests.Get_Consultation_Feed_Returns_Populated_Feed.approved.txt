{
    "reference":"GID-TA10232",
    "title":"For consultation comments",
    "consultationName":"Appraisal consultation",
    "startDate":"2018-04-01T00:00:00",
    "endDate":"2028-05-31T17:00:00",
    "consultationType":"Appraisal consultation",
    "resourceTitleId":null,
    "projectType":"TA",
    "productTypeName":"Technology appraisal guidance",
    "showExpressionOfInterestSubmissionQuestion":true,
    "developedAs":null,
    "relevantTo":null,
    "consultationId":1,
    "process":"TAG",
    "hasDocumentsWhichAllowConsultationComments":false,
    "partiallyUpdatedProjectReference":null,
    "origProjectReference":null,
    "user":{
        "isAuthenticated":true,
        "displayName":"Benjamin Button",
        "userId":"00000000-0000-0000-0000-000000000000",
        "organisationsAssignedAsLead":[
            
        ],
        "validatedSessions":[
            
        ],
        "validatedOrganisationUserIds":[
            
        ]
    },
    "consultationState":{
        "startDate":"2018-04-01T00:00:00",
        "endDate":"2028-05-31T17:00:00",
        "hasQuestions":false,
        "hasUserSuppliedAnswers":false,
        "hasUserSuppliedComments":false,
        "submittedDate":null,
        "documentIdsWhichSupportComments":[
            2,
            2
        ],
        "hasAnyDocumentsSupportingComments":true,
        "consultationIsOpen":true,
        "consultationHasNotStartedYet":false,
        "consultationHasEnded":false,
        "supportsSubmission":false,
        "supportsDownload":false,
        "shouldShowDrawer":true,
        "shouldShowCommentsTab":true,
        "shouldShowQuestionsTab":false
    },
    "breadcrumbs":[
        {
            "label":"Home",
            "url":"/",
            "localRoute":false
        },
        {
            "label":"All consultations",
            "url":"/guidance/inconsultation",
            "localRoute":false
        },
        {
            "label":"Consultation responses",
            "url":"/",
            "localRoute":true
        },
        {
            "label":"For consultation comments",
            "url":"/guidance/indevelopment/gid-ta10232/consultation/",
            "localRoute":false
        }
    ],
    "filters":null
}